### PR TITLE
Fix GAN training.

### LINKED ISSUE #591: GAN training with Pytorch Lightning is broken.
I was trying to train a DCGAN on my dataset but it wouldn't work in any means until I detach the training logic from Lightning and run the code without it. It was not working when my training logic is in Lightning module. I checked the gan examples in the docs and also multiple optimizer things. After 2 days of headaches, source code inspections and putting numerous print statements in the lightning source code, I found the culprit.

GAN training with Pytorch Lightning is simply broken. The culprit is only calling optimizer.zero_grad() after optimizer.step() since it clears the gradients of Generator or Discriminator only. Before the other network's weights are updated for, say, Generator; "loss.backward()" is called and it updates gradients for all parameters, but when optimizer.zero_grad() is called after the parameters are updated, only Generator's gradients are cleaned. So when it comes the Discriminator loss.backward(), leftover gradients are accumulated for Discriminator parameters and it just messes up with everything. Any kind of GAN training is impossible with this settings. That's why you can not find any GAN implementations with Pytorch Lightning on the internet.

Possible solutions
-Putting a warning in the docs or on the console after detecting multiple optimizers are defined like "if you train GAN, don't forget to zero all gradients by overriding the optimizer_step method or just reset gradients in your training loop before returning loss dictionary". (would be weird, honestly)

- calling zero_grad before calling backward in optimizer_closure (but it would mess with gradient accumulation, i suppose)

- calling self.zero_grad() instead of optimizer.zero_grad() by default in the optimizer_step.

- call zero_grad for all optimizers of the lightning module after gradient step by default so that the hook on_before_zero_grad is called whenever appropriate.

(By the way, on_before_zero_grad is not seem to be called anywhere right now. Maybe the issue can be fixed with a default behavior of that method, alternatively.)

I just sent a pull request which implements the last option and updates the docs.

### PR DESCRIPTION
**Pull request by aybberk**
## What does this PR do?
Fixes #591 and #592 in a better way. We can keep the same optimizer_step logic and keep accumulated gradients without calculating unnecessary(and harmful) gradients plus a little speedup.

### CODE REVIEW

**Comment by aybberk** [comment]
Also, a boolean argument in an appropriate method can be added to specify if the user actually wants this feature(preventing gradient calculation other optimizers' parameters) with this default setting. However I just can not think of a case one would want to calculate gradients of parameters that would not be updated right now.

**Comment by williamFalcon** [comment]
wondering if looping over each param for every forward pass will cause a bottleneck

**Comment by aybberk** [comment]
Well at first I suspected it also, but made a naive test to check.

```
                for n in range(100):
                    for param in self.get_model().parameters():
                        param.requires_grad = False
                    for group in optimizer.param_groups:
                        for param in group['params']:
                            param.requires_grad = True
```

When I wrap it in a 100-loop like that, batch/sec does not change on my laptop. (simple DCGAN setup with GTX1060). I'm going to make a little more structured test for that, trying 0, 1, 100, 1000 loops.

**Comment by aybberk** [comment]
0-loop:   20 batch/sec
1-loop:   23.5 batch/sec
10-loop: 23 batch/sec
100-loop: 19.5 batch/sec
1000-loop: 7.5 batch/sec

Seems the looping takes approximately 0.5% of the time passed for a batch (with two optimizers) in the default case.

**Comment by Borda** [comment]
would `map` help? writing something like:
`map(lambda p: p.requires_grad = False, self.get_model().parameters():`

**Comment by aybberk** [comment]
Yeah thanks! It definitely helps. However, lambda expressions does not allow variable assignments. So using https://pytorch.org/docs/stable/tensors.html#torch.Tensor.requires_grad_ would be a better choice. Also, I'm going to try testing list comprehensions also, as we know it's a little faster than map operations https://stackoverflow.com/questions/1247486/list-comprehension-vs-map and I think it is really more readable and aestethic.

**Comment by Borda** [comment]
@aybberk rebase master, pls...

**Comment by aybberk** [comment]
I think I managed to rebase it, but I'm not sure. Could you check it?
Also, do I need to submit a new PR?

Sorry, I'm not really familiar with the sophisticated git features.

Quoting Jirka Borovec <notifications@github.com>:

> @aybberk rebase master, pls...
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/williamFalcon/pytorch-lightning/pull/603#issuecomment-562987717


-- 
Ayberk Aydın

**Comment by jeffling** [comment]
I think this solution captures the core of the problem. Thanks @aybberk!

**Comment by aybberk** [comment]
I think notebooks would not help with reproducibility since I just tweaked the Lightning source code on my machine to make tests. And the tests were nothing too fancy, simply time.time() calls before and after the loop above and also inspecting the batch/sec in the progress bar.

**Comment by williamFalcon** [comment]
@tullie 
thoughts?

**Comment by tullie** [comment]
Looks good to me - this is how i've implemented GANs in the past. 

It can be good to have a helper function for model_requires_grad_on() model_requires_grad_off() because it's a common pattern, but let's leave it for next time it comes up.

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -433,6 +433,13 @@ def run_training_batch(self, batch, batch_idx):
 
             # call training_step once per optimizer
             for opt_idx, optimizer in enumerate(self.optimizers):
+                # make sure only the gradients of the current optimizer's paramaters are calculated 
+                # in the training step to prevent dangling gradients in multiple-optimizer setup.
+                for param in self.get_model().parameters():
+                    param.requires_grad = False
+                for group in optimizer.param_groups:
+                    for param in group['params']:
+                        param.requires_grad = True
```
Comment (CarloLucibello) [inline pytorch_lightning/trainer/training_loop.py]: should all params of all optimizers be set to True after the step in order to restore the original state? I can't think of a concrete example but wonder if this may lead to very hard to find bugs.

Comment (williamFalcon) [reply pytorch_lightning/trainer/training_loop.py]: yeah, i think this should be a freeze, unfreeze type of op. agreed

Comment (williamFalcon) [reply pytorch_lightning/trainer/training_loop.py]: but we really should keep exactly the same state as before. So, blanket freeze unfreeze is bad. we should freeze unfrozen params and unfreeze only those after (keep a dict of indices when freezing, for example)

Comment (aybberk) [reply pytorch_lightning/trainer/training_loop.py]: I also agree, a user may want to inspect gradients sometime in training. Helpers like save_freeze_state() and restore_freeze_state() would definitely help. Also, we only need this logic in multiple optimizer training setup so we can add an "if len(self.optimizers) > 1" to the top so that we can avoid redundantly freezing and unfreezing all parameters at every batch which takes ~0.5% of the time.

**Comment by CarloLucibello** [comment]
the freeze/unfreeze concerns above were not addressed before merge

**Comment by jeffling** [comment]
@CarloLucibello That's a concern. @aybberk do you have time to do your suggested improvements here? https://github.com/PyTorchLightning/pytorch-lightning/pull/603#discussion_r358626166

**Comment by williamFalcon** [comment]
@CarloLucibello totally overlooked this. mind submitting a fix PR?

**Comment by aybberk** [comment]
I am going to submit a fix PR ASAP, possibly tomorrow.

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (kwanUm) [reply pytorch_lightning/trainer/training_loop.py]: @aybberk @williamFalcon - this change perfectly solves any dangling grads issue, but seems to work much slower than the suggested fix at #666 

The change in #666 resets the grads ones - thus communicates with the GPU once at every training step. 
This solution passes on each layer to set require_grads, and if the model runs on a GPU, the GPU will be addressed many times in a row.

From my experiments with this approach, it slows the runtime of the training by a significant factor. Have you benchmarked it yourself? What were your results?

Comment (aybberk) [reply pytorch_lightning/trainer/training_loop.py]: > @aybberk @williamFalcon - this change perfectly solves any dangling grads issue, but seems to work much slower than the suggested fix at #666
> 
> The change in #666 resets the grads ones - thus communicates with the GPU once at every training step.
> This solution passes on each layer to set require_grads, and if the model runs on a GPU, the GPU will be addressed many times in a row.
> 
> From my experiments with this approach, it slows the runtime of the training by a significant factor. Have you benchmarked it yourself? What were your results?

Benchmarks are done in my local pc with gtx1060, and shown a negligible effect(~0.5%) on performance. How was your results? I think multiple sequential GPU communication is not a bottleneck there since when you call a zero_grad() on an optimizer or model, PyTorch does the exact same thing in the lower level, to the best of my information.
